### DISCUSSION TITLE
How to access state values after converting my equatable class to freezed?

### DISCUSSION THREAD

**Discussion by SiroDevs** [rrousselGit/freezed#1010] [Q&A]
I had this state class

```
part of 'login_bloc.dart';

final class LoginState extends Equatable {
  const LoginState({this.username = '', this.password = ''});

  final String username;
  final String password;

  LoginState copyWith({String? username, String? password}) {
    return LoginState(
      username: username ?? this.username,
      password: password ?? this.password,
    );
  }

  @override
  List<Object> get props => [username, password];
}
```

which after conversion according to my research and docs became

```
part of 'login_bloc.dart';

@freezed
abstract class LoginState with _$LoginState {

  const factory LoginState.loginFailure(int timestamp, String code) =
      LoginFailureState;

  const factory LoginState.loginData({String? username, String? password}) = LoginDataState;

  const factory LoginState.loginUsernameChanged(int timestamp, String username) =
      LoginUsernameChangedState;

  const factory LoginState.loginPasswordChanged(
      int timestamp, String password) = LoginPasswordChangedState;

  const factory LoginState.loginSuccess(int timestamp) = LoginSuccessState;

  const factory LoginState.loginSubmitted(int timestamp) = LoginSubmittedState;

}

```

Now in my bloc I had this code before:

  ```
void _onLoginUsernameChanged(
    LoignUsernameChangedEvent event,
    Emitter<LoginState> emit,
  ) {
    emit(
      state.copyWith(username: event.username),
    );
  }
```

Now it is
```
void _onLoginUsernameChanged(
    LoignUsernameChangedEvent event,
    Emitter<LoginState> emit,
  ) {
    emit(
      (state as LoginDataState).copyWith(username: event.username),
    );
  }
```

Now the tricky part is for retrieving the data. Initially when using equatables it was:
```
Future<void> _onLoginSubmitted(
    LoginSubmittedEvent event,
    Emitter<LoginState> emit,
  ) async {
    _authenticationRepository.logIn(
      username: state.username.value,
      password: state.password.value,
    );
  }
```
This is the part I need help with. I need to access data in the LoginDataState and using map is very clear to me on how I can do this.

**Comment by mboyamike** [comment]
For generating the toMap and fromMap methods, you should add the json_serializable - https://pub.dev/packages/json_serializable

Freezed is aware of the package. Add the json_serializable package to your pubspec file, add a part 'login_bloc.g.dart' then add a fromJson method as follows.

```dart
  factory LoginState.fromJson(Map<String, dynamic> json) => _$LoginStateFromJson(json);
```

The caveat is it adds a key method to identify which constructor to call. Something like
{
  key: "loginFailure",
  timestamp: ...,
  code: ...
}

Be aware of this if you get the json from a server that may not have that key

**Reply by mboyamike** [reply]
Also side note, try to avoid tagging Remi in these issues. He still gets to see them with time

Also if you run into issues, a better place for discussion is in the freezed-discussion channel on discord - https://discord.gg/riverpod-765557403865186374

**Reply by SiroDevs** [reply]
Thank you for your answer, unfortunately this didn't help me solve my issue as I expected.
